File: sol/solution/0001-0100/0032. Longest Valid Parentheses/Solution.scala
object Solution {
  def longestValidParentheses(s: String): Int = {
    var ans = 0; var left = 0; var right = 0
    s.foreach { ch =>
      if (ch == '(') left += 1 else right += 1
      if (left == right) ans = math.max(ans, 2 * right)
      else if (right > left) { left = 0; right = 0 }
    }
    left = 0; right = 0
    s.reverse.foreach { ch =>
      if (ch == ')') right += 1 else left += 1
      if (left == right) ans = math.max(ans, 2 * left)
      else if (left > right) { left = 0; right = 0 }
    }
    ans
  }
}

File: sol/solution/0001-0100/0032. Longest Valid Parentheses/Solution.cs
public class Solution {
    public int LongestValidParentheses(string s) {
        int ans = 0, left = 0, right = 0;
        foreach (var c in s) {
            if (c == '(') left++; else right++;
            if (left == right) ans = Math.Max(ans, 2 * right);
            else if (right > left) { left = right = 0; }
        }
        left = right = 0;
        for (int i = s.Length - 1; i >= 0; i--) {
            if (s[i] == ')') right++; else left++;
            if (left == right) ans = Math.Max(ans, 2 * left);
            else if (left > right) { left = right = 0; }
        }
        return ans;
    }
}